src/cli/commands/global.js

```diff
@@ -87,29 +87,31 @@ async function getGlobalPrefix(config: Config, flags: Object): Promise<string> {
     return process.env.PREFIX;
   }
 
-  let prefix = FALLBACK_GLOBAL_PREFIX;
+  const potentialPrefixFolders = [FALLBACK_GLOBAL_PREFIX];
   if (process.platform === 'win32') {
     // %LOCALAPPDATA%\Yarn --> C:\Users\Alice\AppData\Local\Yarn
     if (process.env.LOCALAPPDATA) {
-      prefix = path.join(process.env.LOCALAPPDATA, 'Yarn');
+      potentialPrefixFolders.unshift(path.join(process.env.LOCALAPPDATA, 'Yarn'));
     }
   } else {
-    prefix = POSIX_GLOBAL_PREFIX;
+    potentialPrefixFolders.unshift(POSIX_GLOBAL_PREFIX);
   }
 
-  const binFolder = path.join(prefix, 'bin');
-  try {
-    // eslint-disable-next-line no-bitwise
-    await fs.access(binFolder, fs.constants.W_OK | fs.constants.X_OK);
-  } catch (err) {
-    if (err.code === 'EACCES') {
-      prefix = FALLBACK_GLOBAL_PREFIX;
-    } else if (err.code === 'ENOENT') {
-      // ignore - that just means we don't have the folder, yet
-    } else {
-      throw err;
-    }
+  const binFolders = potentialPrefixFolders.map(prefix => path.join(prefix, 'bin'));
+  const prefixFolderQueryResult = await fs.getFirstSuitableFolder(binFolders);
```

Comment:
Here's the problem with this code:
We only need write permissions when doing `yarn global add` which is usually called with `sudo` or similar. That said if we change our permission checks depending on usage, then we may end up using two different global folders based on how `yarn` is called (with sudo and without) which would make running `yarn global add` useless.

I think we should just check `R_OK` and `X_OK` here but doing that caused some failing tests (which can be fixed, I'm sure). This would also be a breaking change. That said I'm almost certain that `POSIX_GLOBAL_PREFIX/bin` won't be writeable to yarn by default. What do you think?